In [126]:
#importamos los modulos para poder trabajar
import pandas as pd
import os
csvfile = "02-Homework_04-Pandas_Instructions_HeroesOfPymoli_Resources_purchase_data (1).csv"

# Player Count

In [127]:
data= pd.read_csv(csvfile)


In [129]:
#Limpiamos los repetidos
numberPlayers = data["SN"]
cleanNumberPlayers = numberPlayers.drop_duplicates()
totalPlayers = cleanNumberPlayers.count()

d = {'Total PLayer': [totalPlayers]}
d = pd.DataFrame(data=d)
d

Total PLayer
0           576

# Purchasing Analysis (Total)

In [130]:
uniqueItems = data["Item ID"].drop_duplicates().count()
numberPurchases = data["Purchase ID"].drop_duplicates().count()
avgPrice = data["Price"].sum()/numberPurchases
totalPrice = data["Price"].sum()
avgPrice = data["Price"].sum()/numberPurchases
totalPrice = data["Price"].sum()

In [207]:
d = {'Number Unique Items': [uniqueItems],"Average Price":[avgPrice],"Number of Purchases":[numberPurchases],"Total Revenue":[totalPrice]}
d = pd.DataFrame(data=d)

d["Average Price"]=d["Average Price"].map("${:,.2f}".format)
d["Total Revenue"]=d["Total Revenue"].map("${:,.2f}".format)
d

Number Unique Items Average Price  Number of Purchases Total Revenue
0                  179         $3.05                  780     $2,379.77

# Gender Demographics

In [132]:
data= pd.read_csv(csvfile)
dataClean=data.drop_duplicates(["SN"],keep = "last" )
dataClean= dataClean["Gender"].value_counts()
#dataClean= dataClean["Percentage of Players"] = dataClean["Gender"] / newdataAge["Gender"].sum() *100
dataClean = pd.DataFrame(data = dataClean)
dataClean["Percentage of Players"] = dataClean["Gender"] / dataClean["Gender"].sum() *100
dataClean["Percentage of Players"]=dataClean["Percentage of Players"].map("{:.2f}%".format)
dataClean

Gender Percentage of Players
Male                      484                84.03%
Female                     81                14.06%
Other / Non-Disclosed      11                 1.91%

# Purchasing Analysis (Gender)


In [133]:
data= pd.read_csv(csvfile)
purchaseAnalysis= pd.DataFrame(data["Gender"].value_counts())


In [134]:
avgTPPMale = data[data.Gender=="Male"].groupby("SN").sum().Price.mean()
avgTPPFemale = data[data.Gender=="Female"].groupby("SN").sum().Price.mean()
avgTPPOther = data[data.Gender=="Other / Non-Disclosed"].groupby("SN").sum().Price.mean()

d = pd.DataFrame({'Avg Total Purchase per Person': [avgTPPFemale,avgTPPMale,avgTPPOther]},["Female","Male","Other / Non-Disclosed"])


In [135]:
avgMale = data[data.Gender=="Male"].groupby("SN").mean().Price.mean()
avgFemale = data[data.Gender=="Female"].groupby("SN").mean().Price.mean()
avgOther = data[data.Gender=="Other / Non-Disclosed"].groupby("SN").mean().Price.mean()
df = pd.DataFrame({'Average Purchase Price': [avgFemale,avgMale,avgOther]},["Female","Male","Other / Non-Disclosed"])



In [136]:
sumMale = data[data.Gender=="Male"].groupby("SN").sum().Price.sum()
sumFemale = data[data.Gender=="Female"].groupby("SN").sum().Price.sum()
sumOther = data[data.Gender=="Other / Non-Disclosed"].groupby("SN").sum().Price.sum()
dff = pd.DataFrame({'Total Purchase Value': [sumFemale,sumMale,sumOther]},["Female","Male","Other / Non-Disclosed"])


In [206]:
out1 = purchaseAnalysis.join(df,how="outer")
out2 = out1.join(dff,how="outer")
out3 = out2.join(d,how="outer")

out3["Average Purchase Price"]=out3["Average Purchase Price"].map("${:,.2f}".format)
out3["Total Purchase Value"]=out3["Total Purchase Value"].map("${:,.2f}".format)
out3["Avg Total Purchase per Person"]=out3["Avg Total Purchase per Person"].map("${:.2f}".format)
out3


Gender Average Purchase Price Total Purchase Value  \
Female                    113                  $3.19              $361.94   
Male                      652                  $3.01            $1,967.64   
Other / Non-Disclosed      15                  $3.35               $50.19   

                      Avg Total Purchase per Person  
Female                                        $4.47  
Male                                          $4.07  
Other / Non-Disclosed                         $4.56

Age Demographics

In [91]:
#Establishing bins
data= pd.read_csv(csvfile)
bins = [0,9,14,19,24,29,34,39,100]
labels=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

In [92]:
dataClean=data.drop_duplicates(["SN"],keep = "last" )
dataClean["Bucket"] = pd.cut(dataClean["Age"],bins, labels = labels)


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [95]:
dataAge = dataClean.groupby(["Bucket"]).count()
newdataAge = dataAge.drop(columns=["Purchase ID","SN","Age","Gender","Item ID","Item Name"])
newdataAge.columns = ["Total Count"]
newdataAge["Percentage of Players"] = newdataAge["Total Count"] / newdataAge["Total Count"].sum() *100
newdataAge["Percentage of Players"]=newdataAge["Percentage of Players"].map("{:.2f}%".format)
newdataAge

Total Count Percentage of Players
Bucket                                   
<10              17                 2.95%
10-14            22                 3.82%
15-19           107                18.58%
20-24           258                44.79%
25-29            77                13.37%
30-34            52                 9.03%
35-39            31                 5.38%
40+              12                 2.08%

# Purchasing Analysis (Age)

In [208]:
data= pd.read_csv(csvfile)
dataAge2=data
#Establishing bins
bins = [0,9,14,19,24,29,34,39,100]
labels=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

In [211]:
dataAge2["Bucket"] = pd.cut(dataAge2["Age"],bins, labels = labels)
dataAge2.groupby(["Bucket"]).count()

dataAge3 = dataAge2.groupby(["Bucket"]).sum()


In [212]:
contador = dataAge2.groupby(["Bucket"]).count()
dataAge3["Total Count"] = contador["Price"]
dataAge3 = dataAge3.drop(columns=["Purchase ID","Age","Item ID"])
dataAge3["Average Purchase Price"] = dataAge3["Price"] / dataAge3["Total Count"]
dataAge3 = dataAge3[["Total Count","Average Purchase Price","Price"]]


In [213]:
avBucket = []
for i in range(len(labels)):
    avBucket.append(data[data.Bucket==labels[i]].groupby("SN").sum().Price.mean())
avBucket = pd.DataFrame(avBucket)
avBucket.columns = ["Avg Total Purchase per Person"]
avBucket= avBucket.set_index(pd.Index(["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]))


In [214]:
final = dataAge3.join(avBucket, how="outer")

final.rename(columns={"Total Count": "Purchase Count", "Price":"Total Purchase Value"}, inplace = True)


final["Average Purchase Price"]=final["Average Purchase Price"].map("${:,.2f}".format)
final["Total Purchase Value"]=final["Total Purchase Value"].map("${:,.2f}".format)
final["Avg Total Purchase per Person"]=final["Avg Total Purchase per Person"].map("${:,.2f}".format)

final

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

# Top Spenders

In [204]:
data= pd.read_csv(csvfile)


spenders = data.groupby(["SN"])
spenderdf1 = spenders.count()
spenderdf1 = pd.DataFrame(spenderdf1["Purchase ID"])

spenderdf = spenders.sum()
spenderdf = spenderdf.sort_values("Price",ascending = False)
spenderdf = spenderdf.drop(columns = ["Purchase ID","Age","Item ID"])


spenderTotal = pd.merge(spenderdf,spenderdf1,on="SN",how="outer")
spenderTotal["Average Purchase Price"] = spenderTotal["Price"] / spenderTotal["Purchase ID"]
spenderTotal.columns = ["Total Purchase Value","Purchase Count", "Average Purchase Price"]
spenderTotal = spenderTotal[["Purchase Count","Average Purchase Price","Total Purchase Value"]]
spenderTotal["Average Purchase Price"]=spenderTotal["Average Purchase Price"].map("${:,.2f}".format) 
spenderTotal["Total Purchase Value"]=spenderTotal["Total Purchase Value"].map("${:,.2f}".format) 
spenderTotal.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

# Most Popular Items

In [203]:
data= pd.read_csv(csvfile)

item = data.groupby(["Item ID", "Item Name"])
item_agg = item.agg({
    "Price": ["count", "mean", "sum"]
})
item_agg = item_agg.sort_values(('Price', 'count'), ascending = False)


item_agg['Price', 'sum']=item_agg['Price', 'sum'].map("${:,.2f}".format)
item_agg['Price', 'mean']=item_agg['Price', 'mean'].map("${:,.2f}".format)


item_agg.rename(columns={'Price': "" },inplace=True)
item_agg.rename(columns={'count': "Purchase Count" },inplace=True)
item_agg.rename(columns={'mean': "Item Price" },inplace=True)
item_agg.rename(columns={'sum': "Total Purchase Value" },inplace=True)

item_agg.head()

\
                                                     Purchase Count   
Item ID Item Name                                                     
92      Final Critic                                             13   
178     Oathbreaker, Last Hope of the Breaking Storm             12   
145     Fiery Glass Crusader                                      9   
132     Persuasion                                                9   
108     Extraction, Quickblade Of Trembling Hands                 9   

                                                                 \
                                                     Item Price   
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                                           
                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

# Most Profitable Items

In [202]:

data= pd.read_csv(csvfile)

item = data.groupby(["Item ID", "Item Name"])
item_agg2 = item.agg({
    "Price": ["count", "mean", "sum"]
})
item_agg2 = item_agg2.sort_values(('Price', 'sum'), ascending = False).head()

item_agg2['Price', 'sum']=item_agg2['Price', 'sum'].map("${:,.2f}".format)
item_agg2['Price', 'mean']=item_agg2['Price', 'mean'].map("${:,.2f}".format)


item_agg2.rename(columns={'Price': "" },inplace=True)
item_agg2.rename(columns={'count': "Purchase Count" },inplace=True)
item_agg2.rename(columns={'mean': "Item Price" },inplace=True)
item_agg2.rename(columns={'sum': "Total Purchase Value" },inplace=True)

item_agg2

\
                                                     Purchase Count   
Item ID Item Name                                                     
92      Final Critic                                             13   
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
103     Singed Scalpel                                            8   

                                                                 \
                                                     Item Price   
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                                           
                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# You must include a written description of three observable trends based on the data

Most players are male (84%). That means that it is not a very inclusive game/industry. Or that it is not very apealing for the female population. Meaby female players are more interested on other activities.

People more interested in the game are in the 20-24 range. That braket is the one that had the most purchases (365), higest "Total Purchase Value" ($1,114.06) and second higest "Avg Total Purchase per Person" ($4.32)

Eventhough, female population is less represented (14.06%) it has the highest "AVg Total Purchase per Person" meaning that those few female players, really are into gaming